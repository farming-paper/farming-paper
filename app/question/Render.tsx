// Create reference instance
import katex from "katex";
import katexStyle from "katex/dist/katex.min.css";
import { marked } from "marked";
import React from "react";

const renderKatex = (text: string, displayMode: boolean) => {
  return katex.renderToString(text, {
    displayMode,
    macros: {
      "\\RR": "\\mathbb{R}",
      "\\ZZ": "\\mathbb{Z}",
      "\\f": "#1f(#2)",
    },
  });
};

export function links() {
  return [{ rel: "stylesheet", href: katexStyle }];
}

marked.use({
  renderer: {
    paragraph(text) {
      let result = text;
      result = result.replace(/(\$\$((.|\n)+?)\$\$)/gm, (match, p1, p2) => {
        return renderKatex(p2, true);
      });

      result = result.replace(/(\$((.|\n)+?)\$)/gm, (match, p1, p2) => {
        return renderKatex(p2, false);
      });
      return `<p>${result}</p>`;
    },
  },
});

const Render: React.FC<{ children: string }> = ({ children }) => {
  return (
    <div
      className="leading-normal"
      dangerouslySetInnerHTML={{ __html: marked(children) }}
    />
  );
};

export default React.memo(Render);
