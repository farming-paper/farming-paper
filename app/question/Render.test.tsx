import { render } from "@testing-library/react";
import { describe, expect, test } from "vitest";
import Render from "./Render";

test("Simple Text", () => {
  const result = render(<Render>Simple Text</Render>);

  expect(result.container).toMatchSnapshot();
});

describe("마크다운 렌더링", () => {
  test("제목이 잘 표시되어야 함", () => {
    const result = render(
      <Render>{`# heading 1
  
  ## heading 2
  
  ### heading 3`}</Render>
    );

    expect(result.container).toMatchSnapshot();
  });

  test("굵게가 잘 표시되어야 함", () => {
    const result = render(<Render>{`**bold**`}</Render>);
    expect(result.container).toMatchSnapshot();

    // screen.debug();
  });
});

describe("수학 렌더링 (KaTeX)", () => {
  test("Inline Math 가 잘 렌더링되어야 함", () => {
    const result = render(<Render>{`$x^2$`}</Render>);

    expect(result.container).toMatchSnapshot();
  });

  test("Block Math 가 잘 렌더링되어야 함", () => {
    const result = render(<Render>{`$$x^2$$`}</Render>);

    expect(result.container).toMatchSnapshot();
  });
});
