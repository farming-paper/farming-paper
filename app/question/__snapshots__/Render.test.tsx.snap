// Vitest Snapshot v1

exports[`Simple Text 1`] = `
<div>
  <div>
    <p>
      Simple Text
    </p>
    

  </div>
</div>
`;

exports[`마크다운 렌더링 > 굵게가 잘 표시되어야 함 1`] = `
<div>
  <div>
    <p>
      <strong>
        bold
      </strong>
    </p>
    

  </div>
</div>
`;

exports[`마크다운 렌더링 > 제목이 잘 표시되어야 함 1`] = `
<div>
  <div>
    <h1
      id="heading-1"
    >
      heading 1
    </h1>
    

    <h2
      id="heading-2"
    >
      heading 2
    </h2>
    

    <h3
      id="heading-3"
    >
      heading 3
    </h3>
    

  </div>
</div>
`;

exports[`수학 렌더링 (KaTeX) > Block Math 가 잘 렌더링되어야 함 1`] = `
<div>
  <div>
    <p>
      <span
        class="katex-display"
      >
        <span
          class="katex"
        >
          <span
            class="katex-mathml"
          >
            <math
              display="block"
              xmlns="http://www.w3.org/1998/Math/MathML"
            >
              <semantics>
                <mrow>
                  <msup>
                    <mi>
                      x
                    </mi>
                    <mn>
                      2
                    </mn>
                  </msup>
                </mrow>
                <annotation
                  encoding="application/x-tex"
                >
                  x^2
                </annotation>
              </semantics>
            </math>
          </span>
          <span
            aria-hidden="true"
            class="katex-html"
          >
            <span
              class="base"
            >
              <span
                class="strut"
                style="height:0.8641em;"
              />
              <span
                class="mord"
              >
                <span
                  class="mord mathnormal"
                >
                  x
                </span>
                <span
                  class="msupsub"
                >
                  <span
                    class="vlist-t"
                  >
                    <span
                      class="vlist-r"
                    >
                      <span
                        class="vlist"
                        style="height:0.8641em;"
                      >
                        <span
                          style="top:-3.113em;margin-right:0.05em;"
                        >
                          <span
                            class="pstrut"
                            style="height:2.7em;"
                          />
                          <span
                            class="sizing reset-size6 size3 mtight"
                          >
                            <span
                              class="mord mtight"
                            >
                              2
                            </span>
                          </span>
                        </span>
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </span>
        </span>
      </span>
    </p>
    

  </div>
</div>
`;

exports[`수학 렌더링 (KaTeX) > Inline Math 가 잘 렌더링되어야 함 1`] = `
<div>
  <div>
    <p>
      <span
        class="katex"
      >
        <span
          class="katex-mathml"
        >
          <math
            xmlns="http://www.w3.org/1998/Math/MathML"
          >
            <semantics>
              <mrow>
                <msup>
                  <mi>
                    x
                  </mi>
                  <mn>
                    2
                  </mn>
                </msup>
              </mrow>
              <annotation
                encoding="application/x-tex"
              >
                x^2
              </annotation>
            </semantics>
          </math>
        </span>
        <span
          aria-hidden="true"
          class="katex-html"
        >
          <span
            class="base"
          >
            <span
              class="strut"
              style="height:0.8141em;"
            />
            <span
              class="mord"
            >
              <span
                class="mord mathnormal"
              >
                x
              </span>
              <span
                class="msupsub"
              >
                <span
                  class="vlist-t"
                >
                  <span
                    class="vlist-r"
                  >
                    <span
                      class="vlist"
                      style="height:0.8141em;"
                    >
                      <span
                        style="top:-3.063em;margin-right:0.05em;"
                      >
                        <span
                          class="pstrut"
                          style="height:2.7em;"
                        />
                        <span
                          class="sizing reset-size6 size3 mtight"
                        >
                          <span
                            class="mord mtight"
                          >
                            2
                          </span>
                        </span>
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </span>
        </span>
      </span>
    </p>
    

  </div>
</div>
`;
